agent/agent.go

```diff
@@ -0,0 +1,300 @@
+func (a *Agent) Heartbeat() {
+	// Add any heartbeat functionality here.
+}
+
+func (a *Agent) Terminate() {
+	// Add any cleanup/termination functionality here.
+	output.VerbosePrint("[*] Terminating Sandcat Agent... goodbye.")
+}
+
+// Runs a single instruction and send results.
+func (a *Agent) RunInstruction(command map[string]interface{}, payloads []string) {
+	timeout := int(command["timeout"].(float64))
+	result := make(map[string]interface{})
+	commandOutput, status, pid := execute.RunCommand(command["command"].(string), payloads, command["executor"].(string), timeout)
+	result["id"] = command["id"]
+	result["output"] = commandOutput
+	result["status"] = status
+	result["pid"] = pid
+ 	a.beaconContact.SendExecutionResults(a.GetTrimmedProfile(), result)
+}
+
+// Sets the C2 communication channels for the agent according to the specified C2 configuration map.
+// Will default to HTTP if requested C2 is not available or its requirements aren't met. If defaulting
+// to HTTP when it is not available,or if no communication channels are available at all, an error will be returned.
+func (a *Agent) SetCommunicationChannels(c2Config map[string]string) error {
+	if len(contact.CommunicationChannels) > 0 {
+		if c2Name, ok := c2Config["c2Name"]; ok {
+			if requestedComs, ok := contact.CommunicationChannels[c2Name]; ok {
+				if requestedComs.C2RequirementsMet(a.GetFullProfile(), c2Config) {
+					a.beaconContact = requestedComs
+					a.heartbeatContact = requestedComs
+					output.VerbosePrint("[*] Set communication channels for sandcat agent.")
+					return nil
+				} else {
+					output.VerbosePrint("[-] C2 requirements not met! Attempting to default to HTTP.")
+					return a.SetDefaultCommunicationChannel(c2Config)
+				}
+			} else {
+				output.VerbosePrint("[-] Requested C2 channel not found. Attempting to default to HTTP.")
+				return a.SetDefaultCommunicationChannel(c2Config)
+			}
+		} else {
+			output.VerbosePrint("[-] Invalid C2 Configuration. c2Name not specified. Attempting to default to HTTP.")
+			return a.SetDefaultCommunicationChannel(c2Config)
+		}
+	} else {
+		return errors.New("No possible communication channels found.")
+	}
+}
+
+// Sets backup communication channels in case the requested one isn't available.
+// HTTP will be used as the default communication channel. If HTTP isn't available, or if the C2
+// requirements are not met, then an error is returned.
+func (a *Agent) SetDefaultCommunicationChannel(c2Config map[string]string) error {
+	// Default C2 channel is HTTP
+	if coms, ok := contact.CommunicationChannels["HTTP"]; ok {
+		if coms.C2RequirementsMet(a.GetFullProfile(), c2Config) {
+			a.beaconContact = coms
+			a.heartbeatContact = coms
+			output.VerbosePrint("[*] Defaulted to HTTP as communication channels for sandcat agent.")
+			return nil
+		} else {
+			return errors.New("Default HTTP C2 channel requirements not met.")
+		}
+	} else {
+		return errors.New("Default HTTP C2 channel not available.")
+	}
+}
```

Comment:
What is the difference between these? a lot of code re-use too.
```Golang
func (a *Agent) SetCommunicationChannels(c2Config map[string]string) error {
    if len(contact.CommunicationChannels) > 0 {
        selectedC2, ok := c2Config["c2Name"]
        if ok {
            err := attemptSelectChannel(c2Config, selectedC2)
            if err == nil {
                return nil
            }
        }
        err := attemptSelectChannel(c2Config, a.defaultC2)
        if err == nil {
            return nil
        }
    }
    return errors.New("No possible communication channels found.")
} 

func attemptSelectChannel(c2Config map[string]string, commsChannel string) error {
    coms, ok := contact.CommunicationChannels[commsChannel]
    if !ok {
        return errors.New(fmt.Sprintf("%s channel not available", commsChannel)
    }
    if coms.C2RequirementsMet(a.GetFullProfile(), c2Config) {
        a.beaconContact = coms
        a.heartbeatContact = coms
        output.verboseprint(fmt.sprintf("[*] Set communication channel to %s", commsChannel))
        return nil
    }
    return errors.New(fmt.Sprintf("%s channel available, but requirements not met.", commsChannel)
}
```

Reply:
And realistically, that code can be further refactored and refined... it is late though....

Reply:
I like your approach better - definitely cleaner. I'll go ahead and incorporate it in